src/v/archival/service.cc

```diff
@@ -0,0 +1,378 @@
+        // run in parallel
+        ss::gate reconcile_gate;
+        if (!to_remove.empty()) {
+            (void)ss::with_gate(
+              reconcile_gate,
+              [this, to_remove = std::move(to_remove)]() mutable {
+                  return remove_archivers(std::move(to_remove));
+              });
+        }
+        if (!to_create.empty()) {
+            (void)ss::with_gate(
+              reconcile_gate,
+              [this, to_create = std::move(to_create)]() mutable {
+                  return create_archivers(std::move(to_create));
+              });
+        }
+        co_await reconcile_gate.close();
+    }
+    co_return;
+}
+
+ss::future<> scheduler_service_impl::run_uploads() {
+    // Start uploads
+    gate_guard g(_gate);
+    ss::gate upload_gate;
+    ntp_archiver::batch_result total{};
+    int quota = _queue.size();
+    while (_conn_limit.available_units() && quota-- > 0) {
+        auto archiver = _queue.get_upload_candidate();
+        (void)ss::with_gate(
+          upload_gate, [this, archiver, &total]() -> ss::future<> {
+              storage::api& api = _storage_api.local();
+              storage::log_manager& lm = api.log_mgr();
+              vlog(
+                archival_log.trace,
+                "Checking {} for S3 upload candidates",
+                archiver->get_ntp());
+              return archiver->upload_next_candidate(_conn_limit, lm)
+                .then([archiver,
+                       &total](const ntp_archiver::batch_result& res) {
+                    total.num_failed += res.num_failed;
+                    total.num_succeded += res.num_succeded;
+                    if (res.num_failed != 0) {
+                        vlog(
+                          archival_log.error,
+                          "Failed to upload {} segments out of {} that belongs "
+                          "to {}",
+                          res.num_failed,
+                          res.num_succeded,
+                          archiver->get_ntp());
+                    } else if (res.num_succeded) {
+                        vlog(
+                          archival_log.info,
+                          "{} segments that belongs to {} are uploaded",
+                          res.num_succeded,
+                          archiver->get_ntp());
+                    }
+                    return ss::now();
+                });
+          });
+    }
+    co_await upload_gate.close();
+    if (total.num_failed != 0) {
+        vlog(
+          archival_log.error,
+          "Failed to upload {} segments out of {}",
+          total.num_failed,
+          total.num_succeded);
+    }
+    co_return;
+}
+
+ss::future<> scheduler_service_impl::run_cleanup() {
+    // Start uploads
+    gate_guard g(_gate);
+    ss::gate upload_gate;
+    ntp_archiver::batch_result total{};
+    int quota = _queue.size();
+    while (_conn_limit.available_units() && quota-- > 0) {
```

Comment:
same comment here. i wonder if run_xxx should be a background fiber? it would be nice to use the semaphore more traditionally, but it seems to not integrate very nicely with coroutines